README.md
# IPventur-Pro-1.5

famous little IPventur-basic.sh tool, now with menu to choice nmap parameters incl. script, asking for the scan network and the various output files include scanned network and scanned date, IPv6 support

1. download script file IPventur-eng-pro.sh (english version) or IPventur-ger-pro.sh (german version)
2. chmod +x IPventur-eng-pro.sh (or chmod +x IPventur-ger-pro.sh)
3. start script with ./IPventur-eng-pro.sh or when you would like to get the MAC addresses:
   sudo ./IPventur-eng-pro.sh
4. the script is asking for the scanning network now  (don't forget a given subnet mask!)
5. if you would like to use the script parameter, give the correct script name (update your script first with sudo nmap --script-updatedb)
6. check output file e.g.: lanlist-10.0.1.0-08062020.txt  (date format is day month year) and formated file like lanlist-oG-10.0.1.0-08062020.txt or when you used -oA you will get more output files with extentions like .xml, nmap or gnmap.

Feel free to change the nmap script parameters, with your own wishes/needs!


Happy collecting.

MrEnergy64



CHANGELOG.TXT
Version 0.5 origin start script IPventur-eng.sh amd IPventur-ger.sh with standard output and scan and MAC adress

Version 0.6 IPventur-eng-06.sh/IPventur-ger-06.sh - give more comments in the script and extend nmap
            parameters to -A -T5 and add Vendor ID's to the MAC address output
            
Version 1.0 Pro - choice menu for different nmap scan parameters, outputfile includes scanned network and date, minor style changes

Version 1.1 Pro - no start network parameter is needed anymore, script is asking now for the scanning network, minor style changes

Version 1.2 Pro - choice menu for output formats -oN, -oS or -oG, two files are now created (formatted and standard),
            script is now restarting when you pressed wrong keys during the choice menu, minor style changes

Version 1.3 Pro - now with full output format (includes -oA), some grep changes (to the MAC adresse, you will
           get IP and DNS name now) and some minor style changes, change put for formated output to /dev/null, so that the
           output is not in the normal output file and the scan is not visible on the main screen

Version 1.4 Pro - now with IP address/subnet check routine
            - added parameter menu to use scripts
            - some minor style changes

Version 1.4c Pro - moved the fping/nmap check to the beginning
            - added a little bit of color for a better overview
            
Version 1.5 Pro - new design
            - supporting IPv6 now


IPventur-eng-pro.sh
#! /bin/bash
# IPventur.sh Pro
# recommend root rights, need fping and nmap
# updated: 13.06.2020 MrEnergy64 origin: Linux-User
# Version: 1.5
#
clear
# Add network which you like to scan
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-English Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

# check if fping and nmap installed
command -v fping >/dev/null 2>&1 || { echo -e >&2 "I require \e[40m\e[33mfping\e[49m\e[39m but it's not installed.  Please install."; exit 1; }
echo -e "Program \e[40m\e[33mfping\e[49m\e[39m is installed!"
echo
command -v nmap >/dev/null 2>&1 || { echo -e >&2 "I require \e[40m\e[33mnmap\e[49m\e[39m but it's not installed.  Please install."; exit 1; }
echo -e "Program \e[40m\e[33mnmap\e[49m\e[39m  is installed!"
echo ""
echo ""
echo -e "\e[4mWhich network would you like to scan:\e[0m "
echo ""
echo "IPv4 - e.g. 192.168.1.0/24 or 10.0.0.1/32 [Enter] "
echo "IPv6 - e.g. 2a04:35c0:: or 2001:0Db8:85a3:0000:8a2e:0370:7334 [Enter] "
echo ""
read netw
# check if entered IP exist
if [[ -z $netw ]]; then
	clear; echo ""; echo -e "\e[5m no IP address, starting script again....\e[25m ";sleep 4; exec "./IPventur-eng-pro.sh"
fi

# check if entered a valid IPv4 or IPv6 address
if [[ $netw =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\/[0-9]{1,2}$ ]]; then
	clear; echo ""; echo -e " \e[40m\e[33mValid IPv4 address\e[49m\e[39m "

elif [[ $netw =~ ^([0-9a-fA-F]{0,4}:){1,7}[0-9a-fA-F]{0,4}$ ]]; then
       clear; echo ""; echo -e " \e[40m\e[33mValid IPv6 address\e[49m\e[39m "

else
	clear;  echo ""; echo -e "\e[5m $netw is not a valid IPv4 or IPv6 address/subnet mask, starting script again....\e[25m ";sleep 4; exec "./IPventur-eng-pro.sh"
fi

# Menu to choice nmap parameters
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-English Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

echo -e "Scan Network: \e[40m\e[33m$netw\e[49m\e[39m "
echo ""
echo "choice NMAP Scan Version (1,2,3,4,5,6,7 [Enter]): "
echo ""
echo -e "  \e[40m\e[33m1)\e[49m\e[39m NMAP -A 			(intensive scan with OS/Service version 1000 ports), traceroute etc.)"
echo -e "  \e[40m\e[33m2)\e[49m\e[39m NMAP -v -A -p1-65535	(intensive scan with OS/Service version, traceroute etc. more details and all ports)"
echo -e "  \e[40m\e[33m3)\e[49m\e[39m NMAP -6			(standard scan with IPv6)"
echo -e "  \e[40m\e[33m4)\e[49m\e[39m NMAP -F -T5		(fast scan, some standard ports only)"
echo -e "  \e[40m\e[33m5)\e[49m\e[39m NMAP without switches 	(standard scan with open ports)"
echo -e "  \e[40m\e[33m6)\e[49m\e[39m NMAP -d9			(debug scan with highest, could be very long scan!)"
echo -e "  \e[40m\e[33m7)\e[49m\e[39m NMAP -sV --script 		(scan with your entered script)"
echo "				(Check where your scripts are (locate *.nse): ~/.nmap/scripts or /usr/share/nmap/scripts)"
echo "				(Update new NSE NMAP scripts: sudo nmap --script-updatedb)"
echo ""
read n
clear
echo ""
case $n in
# here you can change the namp parameters
  1) echo -e "Choice: \e[40m\e[33mNMAP -A\e[49m\e[39m "; CHOICE="nmap -A";;
  2) echo -e "Choice: \e[40m\e[33mNMAP -v -A -p1-65535\e[49m\e[39m "; CHOICE="nmap -v -A -p1-65535";;
  3) echo -e "Choice: \e[40m\e[33mNMAP -6\e[49m\e[39m "; CHOICE="nmap -6";;
  4) echo -e "Choice: \e[40m\e[33mNMAP -F -R\e[49m\e[39m "; CHOICE="nmap -F -R";;
  5) echo -e "Choice: \e[40m\e[33mNMAP\e[49m\e[39m "; CHOICE="nmap";;
  6) echo -e "Choice: \e[40m\e[33mNMAP -d9\e[49m\e[39m "; CHOICE="nmap -d9";;
  7) echo -e "Choice: \e[40m\e[33m$NMAP -sv --script\e[49m\e[39m "; echo ""; echo -e "Scan Network: \e[40m\e[33m$netw\e[49m\e[39m ";echo "";read -p "Which script would you like to use (type e.g. vulners etc.) => " nms; CHOICE="nmap -sV --script $nms";;
  *) echo ""; echo -e " \e[5minvalide option, starting script again....\e[25m ";sleep 4; exec "./IPventur-eng-pro.sh";;
esac
# Menu to choice output format
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-English Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

echo -e "Scan Network: \e[40m\e[33m$netw\e[49m\e[39m "
echo -e "NMAP Command: \e[40m\e[33m$CHOICE\e[49m\e[39m "
echo ""
echo "choice NMAP output format (1,2,3,4,5 [Enter]): "
echo ""
echo -e "  \e[40m\e[33m1)\e[49m\e[39m -oN 			(Output scan in normal)"
echo -e "  \e[40m\e[33m2)\e[49m\e[39m -oS 			(Output scan in s|<rIpt kIddi3)"
echo -e "  \e[40m\e[33m3)\e[49m\e[39m -oG 			(Output scan in grepable format)"
echo -e "  \e[40m\e[33m4)\e[49m\e[39m -oX			(Output scan in XML format - need to be re-edit)"
echo -e "  \e[40m\e[33m5)\e[49m\e[39m -oA			(Output scan in all formats)"
echo ""
echo "Note: you will get two files (except no. 5): normal output and one whith your choice!"
echo ""
read n
clear
echo ""
case $n in
# here you can change the namp output parameters
  1) echo -e " Choice: \e[40m\e[33m-oN\e[49m\e[39m "; CHOICE2="-oN";;
  2) echo -e " Choice: \e[40m\e[33m-oS\e[49m\e[39m "; CHOICE2="-oS";;
  3) echo -e " Choice: \e[40m\e[33m-oG\e[49m\e[39m "; CHOICE2="-oG";;
  4) echo -e " Choice: \e[40m\e[33m-oX\e[49m\e[39m "; CHOICE2="-oX";;
  5) echo -e " Choice: \e[40m\e[33m-oX\e[49m\e[39m "; CHOICE2="-oA";;
  *) echo ""; echo -e " \e[5minvalide option, starting script again.... \e[25m ";sleep 4; exec "./IPventur-eng-pro.sh";;
esac
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-English Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

echo -e "Scan Network:  \e[40m\e[33m$netw\e[49m\e[39m "
echo -e "NMAP Command:  \e[40m\e[33m$CHOICE\e[49m\e[39m "
echo -e "Output Format: \e[40m\e[33m$CHOICE2\e[49m\e[39m "
echo ""
echo "Overview active network systems:"
echo "(to get MAC addresses, you must be in the same net as root/sudo)"
echo "----------------------------------------------------------------"
echo
date2=$(date +%d.%m.%Y-%H:%M:%S)
date3=$(date +%d%m%Y)
echo -e "Start:   \e[40m\e[33m$date2\e[49m\e[39m "
echo -e "Command: \e[40m\e[33m$CHOICE $CHOICE2 $netw\e[49m\e[39m "
echo
net2=$(echo $netw | cut -d "/" -f 1)
netO=$net2
echo Start: $date2 > lanlist-$netO-$date3.txt
echo Command:  $CHOICE $CHOICE2 $netw >> lanlist-$netO-$date3.txt
echo ""
# Begin creation output file
echo "----------------------------------------------------------------" >> lanlist-$netO-$date3.txt
echo "Network items $date2 / $netw" > lanlist-$netO-$date3.txt
echo "----------------------------------------------------------------" >> lanlist-$netO-$date3.txt
echo "----------------------------------------------------------------"
echo ""  >> lanlist-$netO-$date3.txt
# Scan network and log to the output file, fping checks online IP's only
for k in $(fping -aq -g $netw); do
	echo -e "scanning...: \e[40m$k\e[49m\e[39m "
	echo "Online: $k" >> lanlist-$netO-$date3.txt
# -n Never do DNS resolution resolve, -sP check if IP up and give MAC address with Vendor
	nmap -R -sP $k | awk '/Nmap scan report for/{printf $5" "$6;}/MAC Address:/{print " => "$3" "$4" "$5" "$6;}' | sort >> lanlist-$netO-$date3.txt
# second nmap command with your choiced parameters 
	$CHOICE $k | grep -B1 open >> lanlist-$netO-$date3.txt
# now with your choiced output format
	$CHOICE $k --append-output $CHOICE2 lanlist$CHOICE2-$netO-$date3 >> /dev/null
echo "----------------------------------------------------------------" >> lanlist-$netO-$date3.txt
done
echo "----------------------------------------------------------------"
echo "                            E N D" >> lanlist-$netO-$date3.txt
echo
date2=$(date +%d.%m.%Y-%H:%M:%S)
echo $date2 >> lanlist-$netO-$date3.txt
echo "----------------------------------------------------------------" >> lanlist-$netO-$date3.txt
echo End: $date2
echo 
# Display output file
less lanlist-$netO-$date3.txt


IPventur-ger-pro.sh
#! /bin/bash
# IPventur-Pro.sh
# benötigt root rechte, sowie fping und nmap
# updated: 13.06.2020 MrEnergy64 origin: Linux-User
# Version: 1.5
#
clear
# füge das zu überprüfende Netzwerk hinzu
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-Deutsch Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

# Überprüfe ob fping und nmap vorhanden ist
echo command -v fping >/dev/null 2>&1 || { echo -e >&2 "Programm \e[40m\e[33mfping\e[49m\e[39m ist nicht installiert!  Bitte installieren."; exit 1; }
echo -e "Programm \e[40m\e[33mfping\e[49m\e[39m ist vorhanden!"
echo
command -v nmap >/dev/null 2>&1 || { echo -e >&2 "Programm \e[40m\e[33mnmap\e[49m\e[39m ist nicht installiert!  Bitte installieren."; exit 1; }
echo -e "Programm \e[40m\e[33mnmap\e[49m\e[39m  ist vorhanden!"
echo ""
echo ""
echo -e "\e[4mWelches Netzwerk soll gescannt werden:\e[0m "
echo "IPv4 - z.B. 192.168.1.0/24 oder 192.168.111.1/32 [Enter] "
echo "IPv6 - e.g. 2a04:35c0:: oder 2001:0Db8:85a3:0000:8a2e:0370:7334 [Enter] "
echo ""
read  netw
# checke ob eine IP Adresse eingeben wurde
if [[ -z $netw ]]; then
	clear; echo ""; echo -e "\e[5m Keine IP Adresse vorhanden, starte Script erneut....\e[25m ";sleep 4; exec "./IPventur-ger-pro.sh"
fi

# checke ob die IP Adresse valide ist
if [[ $netw =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\/[0-9]{1,2}$ ]]; then
	clear; echo ""; echo -e " \e[40m\e[33mValide IPv4 Adresse\e[49m\e[39m "

elif [[ $netw =~ ^([0-9a-fA-F]{0,4}:){1,7}[0-9a-fA-F]{0,4}$ ]]; then
       clear; echo ""; echo -e " \e[40m\e[33mValide IPv6 Adresse\e[49m\e[39m "

else
	clear;  echo ""; echo -e "\e[5m $netw ist keine valide IPv4 oder IPv6 Adresse/Subnet mask, starte script erneut....\e[25m ";sleep 4; exec "./IPventur-ger-pro.sh"
fi

# Menü für Parameterübergabe an NMAP
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-Deutsch Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

echo -e "Scanne Netzwerk: \e[40m\e[33m$netw\e[49m\e[39m "
echo ""
echo "wählen Sie eine NMAP Scan Version aus (1,2,3,4,5,6,7 [Enter]): "
echo ""
echo -e "  \e[40m\e[33m1)\e[49m\e[39m NMAP -A 			(intensiver Scan mit OS/Service Version (1000 ports), Traceroute etc.)"
echo -e "  \e[40m\e[33m2)\e[49m\e[39m NMAP -v -A -p1-65535	(größere Ausführlichkeit, intensiver Scan, alle Ports)"
echo -e "  \e[40m\e[33m3)\e[49m\e[39m NMAP -6 			(Standard Scan mit IPv6)"
echo -e "  \e[40m\e[33m4)\e[49m\e[39m NMAP -F -T5 		(schnellster Scan, aber nur Standard Ports)"
echo -e "  \e[40m\e[33m5)\e[49m\e[39m NMAP ohne Parameter 	(Standard Scan mit offenen Ports)"
echo -e "  \e[40m\e[33m6)\e[49m\e[39m NMAP -d9			(Debug mit höster Stufe, kann sehr lange dauern!)"
echo -e "  \e[40m\e[33m7)\e[49m\e[39m NMAP -sV --script 		(Scanne mit Script)"
echo "				(Check wo die Scripts liegen (locate *.nse): ~/.nmap/scripts oder /usr/share/nmap/scripts)"
echo "				(Update neue NSE NMAP Scripts: sudo nmap --script-updatedb)"
echo ""
read n
clear
echo ""
case $n in
# hier können die nmap Parameter geändert werden
  1) echo -e " Auswahl: \e[40m\e[33mNMAP -A\e[49m\e[39m "; AUSWAHL="nmap -A";;
  2) echo -e " Auswahl: \e[40m\e[33mNMAP -v -A -p1-65535\e[49m\e[39m "; AUSWAHL="nmap -v -A -p1-65535";;
  3) echo -e " Auswahl: \e[40m\e[33mNMAP -6\e[49m\e[39m "; AUSWAHL="nmap -6";;
  4) echo -e " Auswahl: \e[40m\e[33mNMAP -F -T5\e[49m\e[39m "; AUSWAHL="nmap -F -T5";;
  5) echo -e " Auswahl: \e[40m\e[33mNMAP\e[49m\e[39m "; AUSWAHL="nmap";;
  6) echo -e " Auswahl: \e[40m\e[33mNMAP -d9\e[49m\e[39m "; AUSWAHL="nmap -d9";;
  7) echo -e " Auswahl: \e[40m\e[33mNMAP -sv --script\e[49m\e[39m "; echo ""; echo -e "Scanne Netzwerk: \e[40m\e[33m$netw\e[49m\e[39m ";echo "";read -p "Whelches Script soll benutzt werden (z.B vulners etc.)?  => " nms; AUSWAHL="nmap -sV --script $nms";;
  *) echo "": echo -e " \e[5mungültige Auswahl, starte Script neu....\e[25m ";sleep 4; exec "./IPventur-ger-pro.sh";;
esac
# Wähle das Ausgabeformat
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-Deutsch Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

echo -e "Scanne Netzwerk: \e[40m\e[33m$netw\e[49m\e[39m"
echo -e "NMAP Parameter:  \e[40m\e[33m$AUSWAHL\e[49m\e[39m"
echo ""
echo "Wähle ein NMAP Ausgabeformat (1,2,3,4,5 [Enter]): "
echo ""
echo -e "  \e[40m\e[33m1)\e[49m\e[39m -oN 			(Ausgabe Scan in normal)"
echo -e "  \e[40m\e[33m2)\e[49m\e[39m -oS 			(Ausgabe Scan in s|<rIpt kIddi3)"
echo -e "  \e[40m\e[33m3)\e[49m\e[39m -oG 			(Ausgabe Scan im Grepable Format)"
echo -e "  \e[40m\e[33m4)\e[49m\e[39m -oX			(Ausgabe Scan im XML Format, muss aber re-editiert werden)"
echo -e "  \e[40m\e[33m5)\e[49m\e[39m -oA			(Ausgabe Scan in allen Formaten)"
echo ""
echo "Hinweis: es werden zwei Ausgabedateien erstellt (Ausnahme Nr.5): eine mit Standard und eine mit Ihrer Wahl!"
echo ""
read n
clear
echo ""
case $n in
# here you can change the namp output parameters
  1) echo -e " \e[40m\e[33mAuswahl: -oN\e[49m\e[39m "; AUSWAHL2="-oN";;
  2) echo -e " \e[40m\e[33mAuswahl: -oS\e[49m\e[39m "; AUSWAHL2="-oS";;
  3) echo -e " \e[40m\e[33mAuswahl: -oG\e[49m\e[39m "; AUSWAHL2="-oG";;
  4) echo -e " \e[40m\e[33mAuswahl: -oX\e[49m\e[39m "; AUSWAHL2="-oX";;
  5) echo -e " \e[40m\e[33mAuswahl: -oA\e[49m\e[39m "; AUSWAHL2="-oA";;
  *) echo ""; echo -e " \e[5mungültige Auswahl, starte Script neu....\e[25m ";sleep 4; exec "./IPventur-ger-pro.sh";;
esac
echo ""
echo "----------------------------------------------------------------"
echo "               ****************************"
echo -e "               * \e[44mIPventur-Deutsch Pro 1.5\e[49m *"
echo "               ****************************"
echo "----------------------------------------------------------------"
echo ""

echo -e "Scanne Netzwerk: 	\e[40m\e[33m$netw\e[49m\e[39m "
echo -e "NMAP Parameter: 	\e[40m\e[33m$AUSWAHL\e[49m\e[39m "
echo -e "Ausgabeformat:		\e[40m\e[33m$AUSWAHL2\e[49m\e[39m "
echo ""
echo "Übersicht aktiver Netzwerkteilnehmer:"
echo "(für MAC Adressen: als root/sudo im selben Netzwerk starten)"
echo "------------------------------------------------------------"
echo
echo
# aktuelles Startdatum und Uhrzeit wird der Ausgangsdatei hinzugefügt
datum=$(date +%d.%m.%Y-%H:%M:%S)
datum2=$(date +%d%m%Y)
echo -e "Start: 	   \e[40m\e[33m$datum\e[49m\e[39m "
echo -e "Kommando:  \e[40m\e[33m$AUSWAHL $AUSWAHL2 $netw\e[49m\e[39m "
netz2=$(echo $netw | cut -d "/" -f 1)
netz=$netz2
echo Start: $datum > lanliste-$netz-$datum2.txt
echo Kommando: $AUSWAHL $AUSWAHL2 $netw >> lanliste-$netz-$datum2.txt
echo ""
# Beginn Ergebnisdatei
echo "------------------------------------------------------------" >> lanliste-$netz-$datum2.txt
echo "Netzwerkbestand $datum / $netw" > lanliste-$netz-$datum2.txt
echo "------------------------------------------------------------" >> lanliste-$netz-$datum2.txt
echo "------------------------------------------------------------"
echo "" >> lanliste-$netz-$datum2.txt

# Scannen des Netztes und Ablage in Ergebnisdatei, fping stellt fest welche IP Online ist
for k in $(fping -aq -g $netw); do
	echo -e "wird untersucht...: \e[40m\e[33m$k\e[49m\e[39m "
	echo "Aktiv: $k" >> lanliste-$netz-$datum2.txt

# -n startet IP DNS Namen check und IP Adresse, -sP zeigt IP Adresse mit MAC Adresse und Hersteller ID
	nmap -R -sP $k | awk '/Nmap scan report for/{printf $5" "$6;}/MAC Address:/{print " => "$3" "$4" "$5" "$6;}' | sort >> lanliste-$netz-$datum2.txt

# -A zeigt mehr Informationen wie OS Version, Traceroute, welche Scripts laufen etc.
	$AUSWAHL $k | grep -B1 open >> lanliste-$netz-$datum2.txt

# zweiter nmap command mit gewähltem Ausgabeformat 
	$AUSWAHL $k --append-output $AUSWAHL2 lanliste$AUSWAHL2-$netz-$datum2 >> lanliste-$netz-$datum2.txt
echo "------------------------------------------------------------" >> lanliste-$netz-$datum2.txt
echo >> lanliste-$netz-$datum2.txt
done
echo "------------------------------------------------------------"
echo "                       E N D E" >> lanliste-$netz-$datum2.txt
echo
datum=$(date +%d.%m.%Y-%H:%M:%S)
echo $datum >> lanliste-$netz-$datum2.txt
echo "------------------------------------------------------------" >> lanliste-$netz-$datum2.txt
echo Ende: $datum
echo
# Anzeige Ergebnisdatei
less lanliste-$netz-$datum2.txt
